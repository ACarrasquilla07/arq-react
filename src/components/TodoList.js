import React, {
 useContext, useEffect, useState, useCallback 
} from 'react';
import { GlobalContext } from '../context/context';
import TodoHeader from './TodoHeader';
import TodoConsumer from '../api/TodoConsumer';

const TodoList = () => {
  const { state, dispatch } = useContext(GlobalContext);
  const [isResult, setIsResult] = useState(null);

  function handleComplete(item) {
    const data = { ...item, complete: true };
    TodoConsumer.update(data, (status) => {
      dispatch({ type: 'COMPLETE', payload: status.item.id });
    });
  }


  useEffect(() => {
    TodoConsumer.getAll((status) => {
      setIsResult(status.isResult);
      if (status.isResult === 'success') {
        dispatch({ type: 'SET_TODO', payload: status.data });
      }
    });
  }, [state.todo]);

  return <>{
    {
      success: <List handleComplete={handleComplete} state={state} />,
      error: <div className="alert alert-dark">Error in the request</div>,
      null: <div>Loading</div>,
    }[isResult]
  }</>;
};

function heander(state) {
  const pluralize = (count) => (count > 1 ? `There are ${count} todos.` : `There is ${count} todo.`);

  return state.todos.length === 0 ? (
    <h4>Yay! All todos are done! Take a rest!</h4>
  ) : (
    <TodoHeader>
        <span className="float-right">{pluralize(state.todos.length)}</span>
      </TodoHeader>
  );
}

const List = ({ state, handleComplete }) => (
<div className="row">
  <div className="col-md-12">
    <div className="row">
      <div className="col-md-12">
        <br />
        {heander(state)}
      </div>
    </div>
    <div className="row">
      <div className="col-md-12">
        <ul className="list-group" data-testid="all-list">
          {state.todos.map(t => (
            <li key={t.id} className="list-group-item" >
              {t.item}
              <button
                className="float-right btn btn-danger btn-sm"
                style={{ marginLeft: 10 }}
                onClick={() => handleComplete(t)}
              >
                Complete
            </button>
            </li>
          ))}
        </ul>
      </div>
    </div>
  </div>
</div>
);

export default TodoList;
