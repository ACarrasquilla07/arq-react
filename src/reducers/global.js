export default function reducer(state, action) {

  if (!action.type) {
    if (action.key instanceof Array) {
      // console.log('state', state.setIn(action.key, action.payload).toJS());
      return state.setIn(action.key, action.payload);
    }
    // console.log('state', state.set(action.key, action.payload).toJS());
    return state.set(action.key, action.payload);
  }

  switch (action.type) {
    case 'ADD_TODO':
      if (!action.payload) {
        return state;
      }
      return {
        ...state,
        todos: [...state.todos, action.payload],
      };
    case 'COMPLETE':
      return {
        ...state,
        todos: state.todos.filter((t) => t.id !== action.payload),
      };
    case 'SET_TODO':
      return {
        ...state,
        todos: action.payload.filter((t) => !t.complete),
      };
    default:
      return state;
  }
}
