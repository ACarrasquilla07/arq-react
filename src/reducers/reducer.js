export default function reducer(state, action) {
  switch (action.type) {
    case 'reset':
      //TODO: descomentar la siguiente linea para resolver el problema del reset
      //return action.payload;
      return null;
    case 'increment':
      return { count: state.count + 1 };
    case 'decrement':
      return { count: state.count - 1 };
    default:
      // A reducer must always return a valid state.
      // Alternatively you can throw an error if an invalid action is dispatched.
      return state;
  }
}