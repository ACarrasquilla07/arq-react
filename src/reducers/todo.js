export default function reducer(state, action) {
  switch (action.type) {
    case "ADD_TODO":
      if (!action.payload) {
        return state;
      }
      return {
        ...state,
        todos: [...state.todos, action.payload]
      };
    case "COMPLETE":
      return {
        ...state,
        todos: state.todos.filter(t => t.id !== action.payload)
      };
    case "SET_TODO":
      return {
        ...state,
        todos: action.payload.filter(t => !t.complete)
      };
    default:
      return state;
  }
}
