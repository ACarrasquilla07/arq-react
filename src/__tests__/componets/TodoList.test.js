import React from "react";
import Enzyme, { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import StoreTodo from "../../context/todo";
import reducers from "../../reducers/todo";
import TodoList from "../../components/TodoList";

Enzyme.configure({ adapter: new Adapter() });

/**
 * usar la libreria Enzyme para buscar los elementos de son cargados
 * @see https://airbnb.io/enzyme/docs/api/ReactWrapper/find.html
 * 
 * como se levanta un contexto de react para poder cargar sus estados??
 * @see https://airbnb.io/enzyme/docs/api/ReactWrapper/mount.html
 * 
 * como se simula los eventos de un elemento
 * @see https://airbnb.io/enzyme/docs/api/ReactWrapper/simulate.html
 */

test("<TodoList /> #display", async () => {
  const todos = ["a", "b", "c"];
  const dispatch = () => {};
  const list = mount(
    <StoreTodo.Provider value={{ state: { todos }, dispatch }}>
      <TodoList />
    </StoreTodo.Provider>
  );

  expect(list.find("li").length).toEqual(3);
  expect(list.find("li").first().html()).toEqual(
    '<li class="list-group-item">a<button class="float-right btn btn-danger btn-sm" style="margin-left: 10px;">Complete</button></li>'
  );
  //TODO: verificar que el ultimo elemento este completado
  //expect(list.find("li").last().....
});

test("<TodoList /> #completeCalls", async () => {
  const todos = ["a", "b", "c"];
  const dispatch = jest.fn();
  const list = mount(
    <StoreTodo.Provider value={{ state: { todos }, dispatch }}>
      <TodoList />
    </StoreTodo.Provider>
  );

   //TODO:  hacer uso del simulador del click, para los 3 casos
   //list.find("button")...

  expect(dispatch.mock.calls.length).toBe(3);
});

test("<TodoList /> #completeMutates", async () => {
  let state = { todos: ["a", "b", "c"] };
  const dispatch = action => {
   /*TODO: usar el reducer. Ej: state =  ????*/
  };
  const list = mount(
    <StoreTodo.Provider value={{ state, dispatch }}>
      <TodoList />
    </StoreTodo.Provider>
  );

  await list
    .find("button")
    .last()
    .simulate("click");

  //TODO: validar para ...toBe(2);
  //TODO: validar para ...toEqual(["a", "b"]);
});
