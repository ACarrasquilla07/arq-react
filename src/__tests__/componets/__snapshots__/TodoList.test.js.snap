// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> #display 1`] = `"<li class=\\"list-group-item\\">task 1<button class=\\"float-right btn btn-danger btn-sm\\" style=\\"margin-left: 10px;\\">Complete</button></li><li class=\\"list-group-item\\">task 2<button class=\\"float-right btn btn-danger btn-sm\\" style=\\"margin-left: 10px;\\">Complete</button></li><li class=\\"list-group-item\\">task 3<button class=\\"float-right btn btn-danger btn-sm\\" style=\\"margin-left: 10px;\\">Complete</button></li>"`;
